### PR TITLE
Fix the mess I introduced ;)

### PR DESCRIPTION
**Pull request by tyx**
getExtensionConfig is in fact not the processed config.
So we do the same that Symfony, parsing config and put expected result in parameter

### CODE REVIEW

**Comment by tyx** [comment]
We still have an issue even with this fix. 

In fact if we config tactician, but not define a command yet. We have the following exception:
> Unable to replace alias "tactician.handler.locator.symfony" with actual definition "tactician.commandbus.default.handler.locator"

I'm on it. And I'm currently working on behats test to be sure everything is really working with Symfony

**Comment by tyx** [comment]
Here is the final PR with all changes required.

- Fix the config
- Assert middleware at compile time (instead of runtime)
- Add functionnal tests

**Comment by tyx** [comment]
And we have a final issue with the doc.

In the doc : 
```yml
tactician:
    commandbus:
        default:
            middleware:
                - tactician.middleware.command_handler
        queued:
            middleware:
                - tactician.middleware.queued_command_handler
```

but with the current code it would be : 
```yml
tactician:
    commandbus:
        default:
            middleware:
                - tactician.middleware.command_handler
        queued:
            middleware:
                - tactician.commandbus.queued.middleware.command_handler
```

Just need to know which one is the right one

**Comment by shouze** [comment]
ping @rosstuck any time to review?

**Comment by rosstuck** [comment]
Honestly, not really, I'm in a huge crunch period right now. I'm trying to enlist some help to assist clear these issues faster. My apologies.

**Comment by tyx** [comment]
Just to let you know, we use this PR in ours projects, everything is working ;)

**Comment by rosstuck** [comment]
My apologies for the inexcusably long delay on this one. While I'm maintaining the other Tactician packages, the Bundle was never something I really worked on or had much experience with. I've tried to fill in as best I can for the other maintainers but clearly, I'm too swamped.

Therefore, the most excellent @frankdejonge has agreed to step in (though where he has the time, I know not) and help out as well. Hopefully this moves your work along in a more timely manner because I'm just up to my eyebrows right now. He'll be at a conference most of this week so it may still take a few days for him to dive in.

My apologies again and thanks for all your hard work.

**Comment by tyx** [comment]
No problem @rosstuck ! 

let me know @frankdejonge if you have any questions ;)

**Comment by RonRademaker** [comment]
Is there any news on this PR?

**Comment by tyx** [comment]
On my side, I just wait for a merge ;)

We use this PR in production on our projects and everything work fine.

File: DependencyInjection/Compiler/UnknownMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace League\Tactician\Bundle\DependencyInjection\Compiler;
+
+use League\Tactician\Bundle\Middleware;
+
+class UnknownMiddleware extends \RuntimeException
```
Comment (RonRademaker) [inline DependencyInjection/Compiler/UnknownMiddleware.php]: To be more explicit I'd suggest renaming this to `UnknownMiddlewareException`

File: DependencyInjection/Compiler/SecurityMiddlewarePass.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (false === $container->hasDefinition('security.authorization_checker')) {
+            return;
+        }
+
+        $container->setDefinition(
+            SecurityMiddleware::SERVICE_ID,
```
Comment (RonRademaker) [inline DependencyInjection/Compiler/SecurityMiddlewarePass.php]: I don't think it's the responsibility of the middleware (not just `SecurityMiddleware` but also the others) to know how there are known in the DIC.

File: TacticianBundle.php [hunk 3]
```diff
@@ -11,8 +10,10 @@ class TacticianBundle extends Bundle
     public function build(ContainerBuilder $container)
     {
         parent::build($container);
-        $container->addCompilerPass(new DoctrineMiddlewarePass());
-        $container->addCompilerPass(new CommandHandlerPass());
+        $container->addCompilerPass(new Compiler\DoctrineMiddlewarePass());
```
Comment (RonRademaker) [inline TacticianBundle.php]: Why did you move the services from `services.yml` to compiler passes?

File: Tests/Bundle/BundleTest.php [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+class BundleTest extends KernelTestCase
+{
+    protected function setUp()
+    {
+        static::$kernel = static::createKernel();
+        $dir = sys_get_temp_dir().DIRECTORY_SEPARATOR.'tactician-bundle'.DIRECTORY_SEPARATOR.md5(microtime(true) * rand(0, 10000));
+        if (false === is_dir($dir)) {
+            mkdir($dir, 0777, true);
+        }
+        static::$kernel->defineCacheDir($dir);
```
Comment (RonRademaker) [inline Tests/Bundle/BundleTest.php]: Perhaps clean this dir after the tests in a `tearDown`?

File: Tests/Bundle/BundleTest.php [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+ *
+ * @runTestsInSeparateProcesses
+ */
+class BundleTest extends KernelTestCase
+{
+    protected function setUp()
+    {
+        static::$kernel = static::createKernel();
+        $dir = sys_get_temp_dir().DIRECTORY_SEPARATOR.'tactician-bundle'.DIRECTORY_SEPARATOR.md5(microtime(true) * rand(0, 10000));
+        if (false === is_dir($dir)) {
```
Comment (RonRademaker) [inline Tests/Bundle/BundleTest.php]: I can't think of an actual case where this dir could possibly exist already 😉

File: Tests/DependencyInjection/Compiler/CommandHandlerPassTest.php [hunk 6]
```diff
@@ -29,16 +29,12 @@ protected function setUp()
 
     public function testProcess()
     {
-        $definition = \Mockery::mock(Definition::class);
-
-        $this->configShouldBe(
+        $this->parametersShouldBe(
```
Comment (RonRademaker) [inline Tests/DependencyInjection/Compiler/CommandHandlerPassTest.php]: Why have these tests been rewritten?

File: DependencyInjection/Compiler/UnknownMiddleware.php (hunk 1 above)
Comment (tyx) [inline DependencyInjection/Compiler/UnknownMiddleware.php]: Matter of taste I guess ;) If it helps to be merged I can change it np.

File: Tests/Bundle/BundleTest.php (hunk 5 above)
Comment (tyx) [inline Tests/Bundle/BundleTest.php]: Indeed :+1:

File: Tests/Bundle/BundleTest.php (hunk 4 above)
Comment (tyx) [inline Tests/Bundle/BundleTest.php]: It helps debug to have it at the end. As we clear on startup, I guess it is enough to prevent weird edge case ?

File: TacticianBundle.php (hunk 3 above)
Comment (tyx) [inline TacticianBundle.php]: Not sure to get the point

File: DependencyInjection/Compiler/SecurityMiddlewarePass.php (hunk 2 above)
Comment (tyx) [inline DependencyInjection/Compiler/SecurityMiddlewarePass.php]: Good point indeed. I'm :+1: 

Let's move the service into the compiler pass ?

File: DependencyInjection/Compiler/UnknownMiddleware.php (hunk 1 above)
Comment (RonRademaker) [inline DependencyInjection/Compiler/UnknownMiddleware.php]: I have no control over merging the PR, I'm just trying to help out by reviewing so the swamped maintainers have less work. Given that the other exceptions in the bundle end with `Exception` (for example `InvalidCommandException`) I think it could help.

File: TacticianBundle.php (hunk 3 above)
Comment (RonRademaker) [inline TacticianBundle.php]: The security- and validator middleware were in services.yml and now they have compiler passes. I was wondering why you decided to change this? 

This is a question without an opinion (at this time_, I'm just curious toward the motivation.

File: Tests/DependencyInjection/Compiler/CommandHandlerPassTest.php (hunk 6 above)
Comment (tyx) [inline Tests/DependencyInjection/Compiler/CommandHandlerPassTest.php]: because they were wrong. They supposed that `getExtensionConfig` returns processed config, it does not in fact.

It is the heart of this PR. We cannot relay on `getExtensionConfig` to get processed config, so we inject stuff from bundle extension into parameters to be able to read it into CompilerPass

File: TacticianBundle.php (hunk 3 above)
Comment (tyx) [inline TacticianBundle.php]: Ok ! Both of them require a suggested dependency as we want the bundle could be used without these middlewares.

Moving to the compiler pass prevents to declare these services that will just not working. Currently we get exception at runtime when trying to use middleware with uninstalled deps. Now we get error at compile time.

**Comment by rosstuck** [comment]
I'm not going to tag a release yet: I'm going to try and write some more integration tests this week, verify this works on a clean setup and give it a once over. Gotta earn my keep sooner or later. ;-)
